Tribler/Policies/BoostingManager.py

```diff
@@ -0,0 +1,676 @@
+                self.torrents[ihash]['enabled'] = mining_bool
+
+                # pause torrent download from disabled source
+                if not mining_bool:
+                    self.stop_download(tor)
+
+        self.boosting_sources[string_to_source(source)].enabled = mining_bool
+
+        self._logger.info("Set mining source %s %s", source, mining_bool)
+
+        if force_restart:
+            self._select_torrent()
+
+    def add_source(self, source):
+        """
+        add new source into the boosting manager
+        """
+        if source not in self.boosting_sources:
+            args = (self.session, source, self.settings, self.on_torrent_insert)
+
+            isdir = os.path.isdir(source)
+
+            if isdir:
+                self.boosting_sources[source] = DirectorySource(*args)
+            elif source.startswith('http://') or source.startswith('https://'):
+                self.boosting_sources[source] = RSSFeedSource(*args)
+            elif len(source) == 20:
+                self.boosting_sources[source] = ChannelSource(*args)
+            else:
+                self._logger.error("Cannot add unknown source %s", source)
+                return
+
+            self._logger.info("Added source %s", source)
+        else:
+            self._logger.info("Already have source %s", source)
+
+    def remove_source(self, source_key):
+        """
+        remove source by stop the downloading and remove its metainfo for all its swarms
+        """
+        if source_key in self.boosting_sources:
+            source = self.boosting_sources.pop(source_key)
+            source.kill_tasks()
+            self._logger.info("Removed source %s", source_key)
+
+            rm_torrents = [torrent for _, torrent in self.torrents.items()
+                           if torrent['source'] == source_to_string(source_key)]
+
+            for torrent in rm_torrents:
+                self.stop_download(torrent)
+                self.torrents.pop(torrent["metainfo"].get_infohash(), None)
+
+            self._logger.info("Torrents download stopped and removed")
+
+    def on_torrent_insert(self, source, infohash, torrent):
+        """
+        This function called when a source finally determined. Fetch some torrents from it,
+        then insert it to our data
+        """
+        def compare_torrents(torrent_1, torrent_2):
+            """
+            comparing swarms. We don't want to download same swarm with different infohash
+            :return: whether those t1 and t2 similar enough
+            """
+            files1 = [files for files in torrent_1['metainfo'].get_files_with_length() if files[1] > 1024 * 1024]
+            files2 = [files for files in torrent_2['metainfo'].get_files_with_length() if files[1] > 1024 * 1024]
+
+            if len(files1) == len(files2):
+                for ft1 in files1:
+                    for ft2 in files2:
+                        if ft1[1] != ft2[1] or lev(ft1[0], ft2[0]) > 5:
```

Comment:
Are the lists sorted on file size at this point? If they are not, this algorithm might report that two torrents are similar while they actually are.